### PR TITLE
Inherent digests

### LINKED ISSUE #2232: Inherent Digests (or, how to get the block author in runtime in a general way)
Oftentimes in a runtime, we want to access the current block author. Right now, the only way that runtimes have to get data like that about the outside world is via inherents. A simple solution would be to add an `Author` inherent and be done with it.

However, for consensus engines we need the property that the header, which doesn't contain any inherents, is enough to verify consensus. Consensus engines like BABE will include the public key of the author in a digest item added _after_ `finalize_block`, along with the signature and other data. Aura is similar, except it uses a slot number instead of a public key, as that is more compact.

Adding an `Author` inherent is still possible, but it comes with a few problems:
  1. Duplication of data when the author would be derivable from the digest seal
  2. potentially complex cross-checking logic to make sure the seal digest item matches the author stated there (usually not a problem)

One possible solution is Inherent Digests. If the import pipeline can be extended to have `initialize_block` take consensus-engine data structured similarly to `InherentData`, this process can initialize the header with digests containing things like `author`.

This would mean that items previously combined in seal digests would be split. The Aura seal can be used as an example. Previously, this would contain the `slot_number` and `signature` of the corresponding public key. Now, the `slot_number` would be passed in as `InherentData` to `initialize_block` and this would perform a runtime lookup to find the public key. A special digest would be produced that encoded the slot number, and the author's key would be available throughout runtime execution (kept in storage, wiped in `on_finalize`). The `signature` would be the only item in the post-runtime digest.

### PR DESCRIPTION
**Pull request by DemiMarie-parity**
Closes #2232

### CODE REVIEW

File: core/consensus/slots/src/lib.rs [hunk 1]
```diff
@@ -193,10 +194,10 @@ pub fn start_slot_worker<B, C, W, T, SO, SC, OnExit>(
 				Ok(Err(())) => warn!(target: "slots", "Authorship task terminated unexpectedly. Restarting"),
 				Err(e) => {
 					if let Some(s) = e.downcast_ref::<&'static str>() {
-						warn!(target: "slots", "Authorship task panicked at {:?}", s);
+						error!(target: "slots", "Authorship task panicked at {:?}", s);
```
Comment (bkchr) [inline core/consensus/slots/src/lib.rs]: Please revert.

File: core/consensus/slots/src/lib.rs [hunk 2]
```diff
@@ -193,10 +194,10 @@ pub fn start_slot_worker<B, C, W, T, SO, SC, OnExit>(
...
 					}
 
-					warn!(target: "slots", "Restarting authorship task");
+					error!(target: "slots", "Restarting authorship task");
```
Comment (bkchr) [inline core/consensus/slots/src/lib.rs]: THis also.

File: core/consensus/slots/src/lib.rs [hunk 3]
```diff
@@ -156,7 +157,7 @@ pub fn start_slot_worker<B, C, W, T, SO, SC, OnExit>(
 
 		// rather than use a timer interval, we schedule our waits ourselves
 		Slots::<SC>::new(slot_duration.slot_duration(), inherent_data_providers)
-			.map_err(|e| debug!(target: "slots", "Faulty timer: {:?}", e))
+			.map_err(|e| error!(target: "slots", "Faulty timer: {:?}", e))
```
Comment (bkchr) [inline core/consensus/slots/src/lib.rs]: Please don't start polluting the logs.

File: core/finality-grandpa/src/lib.rs [hunk 4]
```diff
@@ -51,7 +51,7 @@
 //! number (this is num(signal) + N). When finalizing a block, we either apply
 //! or prune any signaled changes based on whether the signaling block is
 //! included in the newly-finalized chain.
-
+#![forbid(warnings)]
```
Comment (bkchr) [inline core/finality-grandpa/src/lib.rs]:
```suggestion
#![forbid(warnings)]
```

File: core/network/src/test/mod.rs [hunk 5]
```diff
@@ -610,10 +626,14 @@ pub trait TestNetFactory: Sized {
 
 	/// Create new test network with this many peers.
 	fn new(n: usize) -> Self {
+		trace!(target: "test_network", "Creating test network");
```
Comment (bkchr) [inline core/network/src/test/mod.rs]: Why all this spam?

File: core/primitives/src/crypto.rs [hunk 6]
```diff
@@ -288,8 +288,9 @@ impl<T: AsMut<[u8]> + AsRef<[u8]> + Default + Derive> Ss58Codec for T {
 ///
 /// For now it just specifies how to create a key from a phrase and derivation path.
 #[cfg(feature = "std")]
-pub trait Pair: Sized {
+pub trait Pair: Sized + 'static {
 	/// TThe type which is used to encode a public key.
+	// type Public: Send + Sync + Hash + Eq + Clone + Debug + Encode + Decode + 'static;
```
Comment (bkchr) [inline core/primitives/src/crypto.rs]: Why this comment?

File: core/primitives/src/crypto.rs [hunk 7]
```diff
@@ -298,6 +299,7 @@ pub trait Pair: Sized {
 
 	/// The type used to represent a signature. Can be created from a key pair and a message
 	/// and verified with the message and a public key.
+	// type Signature: Send + Sync + Hash + Eq + Clone + Debug + Encode + Decode + 'static;
```
Comment (bkchr) [inline core/primitives/src/crypto.rs]: Same.

File: srml/babe/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,155 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
```
Comment (bkchr) [inline srml/babe/src/lib.rs]:
```suggestion
// Copyright 2019 Parity Technologies (UK) Ltd.
```

File: srml/babe/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+	fn error_to_string(&self, error: &[u8]) -> Option<String> {
+		RuntimeString::decode(&mut &error[..]).map(Into::into)
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Babe {
+		// The last timestamp.
+		LastTimestamp get(last) build(|_| T::Moment::sa(0)): T::Moment;
```
Comment (bkchr) [inline srml/babe/src/lib.rs]:
```suggestion
		LastTimestamp get(last): T::Moment;
```

**Comment by bkchr** [comment]
Is this in progress? Maybe next time open it as a draft.

**Comment by DemiMarie-parity** [comment]
> Is this in progress? Maybe next time open it as a draft.

It is.  I am not too familiar with draft PRs.

File: core/consensus/babe/src/lib.rs [hunk 10]
```diff
@@ -75,28 +155,920 @@ impl Decode for BabeSeal {
...
+fn author_block(
+	randomness: &[u8],
+	slot_number: u64,
+	genesis_hash: &[u8],
+	epoch: u64,
+	authorities: &[sr25519::Public],
+	key: &sr25519::Pair,
+	threshold: u64,
+) -> Option<(VRFInOut, VRFProof, VRFProofBatchable)> {
+	// FIXME this is O(n)
```
Comment (joepetrowski) [inline core/consensus/babe/src/lib.rs]: Issue number :)

File: core/consensus/common/src/block_import.rs [hunk 11]
```diff
@@ -90,6 +90,24 @@ pub enum ForkChoiceStrategy {
 	Custom(bool),
 }
 
+/// The consensus pre-digests.  These are passed to the runtime, but are
```
Comment (joepetrowski) [inline core/consensus/common/src/block_import.rs]: One space between sentences

File: core/consensus/slots/src/lib.rs [hunk 12]
```diff
@@ -237,7 +238,25 @@ impl SlotData for u64 {
...
+
+impl<T> Deref for SlotDuration<T> {
+	type Target = T;
+    fn deref(&self) -> &T {
+		&self.0
+	}
+}
+
+impl<T: SlotData + Clone> SlotData for SlotDuration<T> {
+	/// Get the slot duration in milliseconds
```
Comment (joepetrowski) [inline core/consensus/slots/src/lib.rs]:
```suggestion
	/// Get the slot duration in milliseconds.
```

File: core/consensus/babe/src/lib.rs [hunk 13]
```diff
@@ -75,28 +155,920 @@ impl Decode for BabeSeal {
...
+/// A slot duration. Create with `get_or_compute`.
+// FIXME: Once Rust has higher-kinded types, the duplication between this
+// and `super::babe::Config` can be eliminated.
+pub struct Config(slots::SlotDuration<BabeConfiguration>);
+
+impl Config {
+	/// Either fetch the slot duration from disk or compute it from the genesis
+	/// state.
+	pub fn get_or_compute<B: Block, C>(client: &C) -> CResult<Self>
+	where
```
Comment (joepetrowski) [inline core/consensus/babe/src/lib.rs]: Style: I think this `where` can be bumped up to the previous line or should be indented.

File: core/consensus/slots/src/slots.rs [hunk 14]
```diff
@@ -38,6 +38,11 @@ pub fn duration_now() -> Option<Duration> {
 	}).ok()
 }
 
+/// Get the slot for now.
+pub fn slot_now(slot_duration: u64) -> Option<u64> {
+	duration_now().map(|s| s.as_secs() / slot_duration)
```
Comment (joepetrowski) [inline core/consensus/slots/src/slots.rs]: The struct `SlotDuration` specifies time in milliseconds, is there a unit error here? If you are expecting seconds as the `slot_duration` input, I think the doc comment should specify that it expects an input in seconds.

File: core/network/src/test/mod.rs [hunk 15]
```diff
@@ -165,8 +169,14 @@ impl<S: NetworkSpecialization<Block>> Link<Block> for TestLink<S> {
 		self.link.restart();
 	}
 
+	/// Send synchronization request to the block import channel.
+	///
+	/// The caller should wait for Link::synchronized() call to ensure that it has synchronized
```
Comment (joepetrowski) [inline core/network/src/test/mod.rs]:
```suggestion
	/// The caller should wait for the `Link::synchronized` call to ensure that it has synchronized
```

File: core/network/src/test/mod.rs [hunk 16]
```diff
@@ -165,8 +169,14 @@ impl<S: NetworkSpecialization<Block>> Link<Block> for TestLink<S> {
...
+	/// with ImportQueue.
```
Comment (joepetrowski) [inline core/network/src/test/mod.rs]:
```suggestion
	/// with `ImportQueue`.
```

File: core/network/src/test/mod.rs (hunk 15 above)
Comment (joepetrowski) [reply core/network/src/test/mod.rs]: Actually, as these are tests, do we even need `///` comments?

File: srml/babe/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Consensus extension module for Babe consensus.
```
Comment (joepetrowski) [inline srml/babe/src/lib.rs]:
```suggestion
//! Consensus extension module for BABE block production.
```

File: srml/babe/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+use timestamp::TimestampInherentData;
+use parity_codec::Decode;
+use inherents::{RuntimeString, InherentIdentifier, InherentData, ProvideInherent, MakeFatalError};
+#[cfg(feature = "std")]
+use inherents::{InherentDataProviders, ProvideInherentData};
+
+/// The babe inherent identifier.
+pub const INHERENT_IDENTIFIER: InherentIdentifier = *b"babeslot";
+
+/// The type of the babe inherent.
```
Comment (joepetrowski) [inline srml/babe/src/lib.rs]:
```suggestion
/// The type of the BABE inherent.
```

File: srml/babe/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+pub type InherentType = u64;
+
+/// Auxiliary trait to extract babe inherent data.
```
Comment (joepetrowski) [inline srml/babe/src/lib.rs]:
```suggestion
/// Auxiliary trait to extract BABE inherent data.
```

File: srml/babe/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+pub trait BabeInherentData {
+	/// Get babe inherent data.
```
Comment (joepetrowski) [inline srml/babe/src/lib.rs]:
```suggestion
	/// Get BABE inherent data.
```

File: srml/babe/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+	fn babe_inherent_data(&self) -> result::Result<InherentType, RuntimeString>;
+	/// Replace babe inherent data.
```
Comment (joepetrowski) [inline srml/babe/src/lib.rs]:
```suggestion
	/// Replace BABE inherent data.
```

File: srml/babe/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+		LastTimestamp get(last): T::Moment;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin { }
+}
+
+impl<T: Trait> Module<T> {
+	/// Determine the Babe slot-duration based on the timestamp module configuration.
```
Comment (joepetrowski) [inline srml/babe/src/lib.rs]:
```suggestion
	/// Determine the BABE slot duration based on the Timestamp module configuration.
```

File: core/network/src/test/mod.rs (hunk 5 above)
Comment (DemiMarie-parity) [reply core/network/src/test/mod.rs]: I used it for debugging BABE.  This branch is based on that one.

File: core/consensus/babe/primitives/src/lib.rs [hunk 23]
```diff
@@ -49,6 +64,14 @@ impl BabeConfiguration {
 	pub fn slot_duration(&self) -> u64 {
 		self.slot_duration
 	}
+
+	/// The maximum permitted VRF output, or *threshold*, for BABE.  Currently,
```
[the remark this answers is not in the document]
Comment (DemiMarie-parity) [reply core/consensus/babe/primitives/src/lib.rs]: I removed the trivial getters, since the fields are public.  This is not Java :smile:

**Comment by DemiMarie-parity** [comment]
Closing in favor of #2466

**Pull request closed without merging**